File: HTTPShortcuts/app/src/main/java/ch/rmy/android/http_shortcuts/import_export/ImportTask.java
package ch.rmy.android.http_shortcuts.import_export;

import android.content.Context;
import android.net.Uri;
import android.view.View;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;

import ch.rmy.android.http_shortcuts.R;
import ch.rmy.android.http_shortcuts.realm.Controller;
import ch.rmy.android.http_shortcuts.realm.models.Base;
import ch.rmy.android.http_shortcuts.utils.GsonUtil;
import ch.rmy.android.http_shortcuts.utils.LauncherShortcutManager;

public class ImportTask extends SimpleTask<Uri> {

    public ImportTask(Context context, View baseView) {
        super(context, baseView);
    }

    @Override
    protected Boolean doInBackground(Uri... uris) {
        Uri uri = uris[0];

        Controller controller = null;
        Reader reader = null;
        try {
            try {
                controller = new Controller();
                InputStream inputStream = getContext().getContentResolver().openInputStream(uri);
                if (inputStream == null) {
                    return false;
                }
                reader = new BufferedReader(new InputStreamReader(inputStream));
                Base base = GsonUtil.importData(reader);
                ImportMigrator.migrate(base);
                controller.importBase(base);
                LauncherShortcutManager.updateAppShortcuts(getContext(), controller.getCategories());
            } finally {
                if (reader != null) {
                    reader.close();
                }
                if (controller != null) {
                    controller.destroy();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
        return true;
    }

    @Override
    protected String getProgressMessage() {
        return getString(R.string.import_in_progress);
    }

    @Override
    protected String getSuccessMessage() {
        return getString(R.string.import_success);
    }

    @Override
    protected String getFailureMessage() {
        return getString(R.string.import_failed);
    }

}

File: HTTPShortcuts/app/src/main/kotlin/ch/rmy/android/http_shortcuts/import_export/ImportTask.kt
package ch.rmy.android.http_shortcuts.import_export

import android.content.Context
import android.net.Uri
import android.view.View
import ch.rmy.android.http_shortcuts.R
import ch.rmy.android.http_shortcuts.realm.Controller
import ch.rmy.android.http_shortcuts.utils.GsonUtil
import ch.rmy.android.http_shortcuts.utils.LauncherShortcutManager
import com.google.gson.JsonParser
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class ImportTask(context: Context, baseView: View) : SimpleTask<Uri>(context, baseView) {

    override fun doInBackground(vararg uris: Uri): Exception? {
        val uri = uris[0]

        Controller().use { controller ->
            return try {
                val inputStream = context.contentResolver.openInputStream(uri)
                    ?: return IOException("Failed to open input stream")
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    val importData = JsonParser().parse(reader)

                    val migratedImportData = ImportMigrator.migrate(importData)
                    val newBase = GsonUtil.importData(migratedImportData)
                    controller.importBaseSynchronously(newBase)
                    LauncherShortcutManager.updateAppShortcuts(context, controller.getCategories())
                }
                null
            } catch (e: Exception) {
                e
            }
        }
    }

    override val progressMessage = getString(R.string.import_in_progress)

    override val successMessage = getString(R.string.import_success)

    override val failureMessage = getString(R.string.import_failed)

}